experiments/crossing_validation.py

```diff
@@ -0,0 +1,130 @@
+from algorithm.data_struct import *
+from algorithm.tree import build_tree, c45
+import json
+import random
+import matplotlib.pyplot as plt
+
+def test_crossing_validation(acteptors_examples: [Example], donors_examples: [Example]):
+    k = 10
+    iteration = 100
+    acteptors_results = []
+    donors_results = []
+    acteptors_results_id3 = []
+    donors_results_id3 = []
+    
+    a_result = 0
+    d_result = 0
+    a_result_id3 = 0
+    d_result_id3 = 0
+    for i in range(0, iteration):
+        a_result += crossing_validation(acteptors_examples, k)
+        d_result += crossing_validation(donors_examples, k)
+        a_result_id3 += crossing_validation(acteptors_examples, k, "id3")
+        d_result_id3 += crossing_validation(donors_examples, k, "id3")
+        
+    acteptors_results.append(((str)(k), a_result/iteration))
+    donors_results.append(((str)(k), d_result/iteration))
+    acteptors_results_id3.append(((str)(k), a_result_id3/iteration))
+    donors_results_id3.append(((str)(k), d_result_id3/iteration))
+    
+    save_results(acteptors_results, "c45", "acteptors")
+    save_results(donors_results, "c45", "donors")
+    #file_name = "results/c45_aceptors_crossing_validation.json"
+    #with open(file_name, "w") as file:
+    #    json.dump(acteptors_results, file)
+    #file_name = "results/c45_donors_crossing_validation.json"
+    #with open(file_name, "w") as file:
+    #    json.dump(donors_results, file)
+    
+    #for result in acteptors_results:
+    #    print("Average C4.5 acteptors matching for a number of subsets of  " + str(result[0]) + ": " + str(result[1]))
+    #print()
+    #for result in donors_results:
+    #    print("Average C4.5 donors matching for a number of subsets of " + str(result[0]) + ": " + str(result[1]))
+    #print()
+    
+    save_results(acteptors_results_id3, "id3", "acteptors")
+    save_results(donors_results_id3, "id3", "donors")
+    #file_name = "results/id3_aceptors_crossing_validation.json"
+    #with open(file_name, "w") as file:
+    #    json.dump(acteptors_results_id3, file)
+    #file_name = "results/id3_donors_crossing_validation.json"
+    #with open(file_name, "w") as file:
+    #    json.dump(donors_results_id3, file)
+    
+    #for result in acteptors_results_id3:
+    #    print("Average id3 acteptors matching for a number of subsets of " + str(result[0]) + ": " + str(result[1]))
+    #print()
+    #for result in donors_results_id3:
+    #    print("Average id3 donors matching for a number of subsets of " + str(result[0]) + ": " + str(result[1]))
+
+    save_plot("acteptors_matching", acteptors_results, acteptors_results_id3)
+    save_plot("donors_matching", donors_results, donors_results_id3)
+    
+
+def crossing_validation(examples: [Example], k: int, tree_type = "C4.5"):
```

Comment:
to już czysto programistyczna uwaga, ale taki argument tree_type jest słaby. Chyba najlepiej było by zrobić najpierw dwie metody `build_tree` i `build_tree_c45` następnie analogicznie `crossing_validation_id3` i `crossing_validation_c45` które korzystają z `crossing_validation` które przyjmuje jako argument tree_bulider jako funkcje odpowiednią (`build_tree` lub `build_tree_c45`). No ale to już jak chcesz tak też działa a w ramach projektu może być wyjebka xD

Reply:
w jaki sposób chcesz przekazywać ten tree_builder? W sensie czym dokładnie miałby być?

Reply:
Funkcja która z examples robi tree. Możesz przekazać funkcje jak coś to poczytaj w necie